---
title: "DNAM1_manuscript_Analysis"
author: "Dillon Corvino"
date: "11/06/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console

require(rstudioapi)
require(ggplot2)
require(gplots)
require(TCGAbiolinks)
require(SummarizedExperiment)
require(car)
require(psych)


#Set wd to source file location
setwd(dirname(getActiveDocumentContext()$path))
getwd()


#source("https://bioconductor.org/biocLite.R")
#biocLite("TCGAbiolinks")

# If GDC server down, may need to update TCGAbiolinks, run the below code and RESTART R
#devtools::install_github(repo = "BioinformaticsFMRP/TCGAbiolinks",dependencies = T)

#source('https://bioconductor.org/biocLite.R')
#BiocManager::install("remotes")
#BiocManager::install('Bioconductor/GenomicDataCommons')
#BiocManager::install('BioinformaticsFMRP/TCGAbiolinks',ref = "GenomicDataCommons")

library("GenomicDataCommons")

GenomicDataCommons::status()


if(!dir.exists("Output")){dir.create("Output")}
if(!dir.exists("Output/Data")){dir.create("Output/Data")}

```

```{r Moving_average_analysis}

#############################################
# Moving average analayis was adapted from:
# Riesenberg, S. et al. MITF and c-Jun antagonism interconnects melanoma dedifferentiation with pro-inflammatory cytokine responsiveness and myeloid cell recruitment. Nat. Commun. 6:8755 doi: 10.1038/ncomms9755 (2015).
#############################################

# Required packages for Moving average analysis 
library("Biobase")
library("cgdsr")


# Which TCGA dataset to assay
tcga.cancer <- "skcm_tcga" # or hnsc_tcga etc...


# Create output directory
output.dir <- paste0("Output/Moving_average_", tcga.cancer)

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir))}


########################################
# Function for moving average analysis
########################################
#  see also http://www.cookbook-r.com/
#  free usage under creative commons licence http://creativecommons.org/licenses/by-sa/3.0/legalcode

movingAverage <- function(x, n=1, centered=FALSE) {
  
  if (centered) {
    before <- floor  ((n-1)/2)
    after  <- ceiling((n-1)/2)
  } else {
    before <- n-1
    after  <- 0
  }
  
  # Track the sum and count of number of non-NA items
  s     <- rep(0, length(x))
  count <- rep(0, length(x))
  
  # Add the centered data 
  new <- x
  # Add to count list wherever there isn't a 
  count <- count + !is.na(new)
  # Now replace NA_s with 0_s and add to total
  new[is.na(new)] <- 0
  s <- s + new
  
  # Add the data from before
  i <- 1
  while (i <= before) {
    # This is the vector with offset values to add
    new   <- c(rep(NA, i), x[1:(length(x)-i)])
    
    count <- count + !is.na(new)
    new[is.na(new)] <- 0
    s <- s + new
    
    i <- i+1
  }
  
  # Add the data from after
  i <- 1
  while (i <= after) {
    # This is the vector with offset values to add
    new   <- c(x[(i+1):length(x)], rep(NA, i))
    
    count <- count + !is.na(new)
    new[is.na(new)] <- 0
    s <- s + new
    
    i <- i+1
  }
  
  # return sum divided by count
  s/count
}


#############################################
# Get tcga data using CGDS R package
#############################################
# see also cBioportal http://www.cbioportal.org/cgds_r.jsp
 
# create CGDS object and connection to cBioportal
mycgds = CGDS("http://www.cbioportal.org/", verbose = T)

# test Bioportal connection
test(mycgds)

# check list of cancer studies at server 
getCancerStudies(mycgds)[,1:2]

# check status and available data
getGeneticProfiles(mycgds, tcga.cancer)[,1:2]
getCaseLists(mycgds, tcga.cancer)[,1:2]

#############################################
# Get gene expression data using getProfileData
#############################################
genetic_profile_id <-paste(tcga.cancer,"_rna_seq_v2_mrna", sep ="")
genetic_profile_id

case_list_id <- paste(tcga.cancer, "_rna_seq_v2_mrna", sep = "")
case_list_id


#############################################
# enter genes of interest to retrieve via cBioportal
#############################################

# Reading in GOI
genes.of.interest <- read.table("Genes_of_interest.txt", header = TRUE)
genes.to.plot <- as.character(genes.of.interest$Ensemble)
names(genes.to.plot) <- as.character(genes.of.interest$GeneID)

# Append DNAM1 (CD226) to list of GOIs
gene <- c(paste(c("CD226", names(genes.to.plot))))
length(gene) # 22 genes

# get gene expression data frame 
gedf <- getProfileData(mycgds, gene, genetic_profile_id, case_list_id)


# check normalized gene expression data (read counts) and log2 transform (avoid negative values)

dim(gedf)
ncol(gedf) == length(gene) # TRUE
head(gedf)
gedf[gedf < 1] <- 1
gedf <- log2(gedf)
head(gedf)
n.val <- nrow(gedf)


# define gene used for ranking of samples
xSortGene <- gedf$CD226
xSortGeneID <- "CD226"

xsort <- sort(xSortGene, decreasing = FALSE, index.return = TRUE)


for(gene.i in 2:length(gene)){

  # define dependent gene of interest
  xDepGene <- gedf[ , colnames(gedf) == gene[gene.i]]
  xDepGeneID <- gene[gene.i]
  
  xLeftGene <- xDepGene[xsort$ix] - median(xDepGene[xsort$ix])
  
  
  par(mar = c(5.1, 4.1, 4.1, 4.2))
  
  # Set plot params
  sample.width = 20 # change sample window size for moving average calculation
  laxis.col = "blue" # change colour of dependent gene
  x <- xLeftGene
  
  ylim.plot <- c(summary(xLeftGene)[1], summary(xLeftGene)[6]) # change scale of dependent gene y-axis
  
  # Plot bars of dependent gene expression as a background trace
  plot(x, col = "lightgrey", type = "h", lwd = 0.5, ylim = ylim.plot, xlab = "", ylab ="", axes = F)
  box(which = "plot")
  axis(side = 1, outer = F)
  
  # Plot moving average line for the dependent gene
  axis(side = 2, col = laxis.col, col.axis = laxis.col, lwd = 2)
  abline(h = 0, col = "lightgrey", lwd = 0.5)
  lines(movingAverage(x, n = sample.width, centered =T), col = laxis.col, type = "l", lwd = 3)
  par(new=T)
  
  # Plot line for the ranking gene 
  plot(xsort$x - median(xsort$x), col = "black", type = "l", lwd = 3, xlab = "", ylab ="", axes = F)
  axis(side = 4, col = "black", col.axis = "black", lwd = 2)
  
  # Add aces labels
  title(main = xDepGeneID,
        col.main = laxis.col,
        xlab = "TCGA samples ranked by CD226 expression",
        ylab = paste0("Moving average ", xDepGeneID),
        col.lab = laxis.col)
  
  mtext(paste0(xSortGeneID, " expression (log2)"), side = 4, line = 2.3)
  
  
  # Correlation values
  cor.val <- cor(xLeftGene, xsort$x, method = "spearman")
  cor.val <- signif(cor.val, 4)
  mtext(paste0("r = ", cor.val), adj = 0)
  
  
  cor.test.val <- cor.test(xLeftGene, xsort$x, method = "spearman")
  corr.pval <- signif(cor.test.val$p.value, 4)
  mtext(paste0("P = ", corr.pval), adj = 1)
  
  mtext(paste0("n = ", n.val), adj = 0.5)
  
  # Save plot to file
  dev.copy(pdf, paste0(output.dir, "/Moving_average_CD226_vs_", xDepGeneID, ".pdf"))
  dev.off()

}

```

```{r DNAM1_TCGA_analysis}

########
# DNAM1
########

# Source functions 
source("Functions.R")

# preparing gene of interest vector
genes.of.interest.Ensembl <- read.table("Genes_of_interest.txt", header = T)
goi.plot.vector <- as.character(genes.of.interest.Ensembl$Ensemble)
names(goi.plot.vector) <- as.character(genes.of.interest.Ensembl$GeneID)


#data.sets.vector <- c("TCGA-READ", "TCGA-THCA", "TCGA-MESO", "TCGA-SARC", "TCGA-LGG", "TCGA-ACC", "TCGA-OV", #"TCGA-CHOL", "TCGA-SKCM", "TCGA-GBM", "TCGA-KIRC", "TCGA-BRCA", "TCGA-UCEC", "TCGA-PRAD", "TCGA-LAML", "TCGA-STAD", #"TCGA-LUSC", "TCGA-KICH", "TCGA-TGCT", "TCGA-DLBC", "TCGA-THYM", "TCGA-UVM", "TCGA-HNSC", "TCGA-ESCA", "TCGA-COAD", #"TCGA-LIHC", "TCGA-CESC", "TCGA-KIRP", "TCGA-PAAD", "TCGA-PCPG", "TCGA-UCS", "TCGA-LUAD", "TCGA-BLCA")

data.sets.vector <- c("TCGA-SKCM", "TCGA-HNSC")


for(i in 1:length(data.sets.vector)){
  require(rstudioapi)
  
  if(i == 1){
    prim.dir <- dirname(getActiveDocumentContext()$path)
    print(paste0("Primary Dir = ", getwd()))
    }
  
   
  setwd(prim.dir)
  
  # Listing folders in data
  dir.list <- list.dirs("Data/")
  folder.name.vec <- gsub("TCGA-", "", data.sets.vector)
  
  # Matching folder name with dataset name
  folder.logic <- grepl(folder.name.vec[i], dir.list)
  data.path <- dir.list[folder.logic]
  
  if(sum(folder.logic) == 0){
    print(paste0("Data Folder for ", data.sets.vector[i], " doesn't exist"))
    next
  }
  
  # Moving into dataset folder
  setwd(dir.list[folder.logic])
  
  # Listing files in dataset folder
  file.list <- list.files()
  
  # searching for raw_counts.rda file for dataset
  pattern <- paste0(data.sets.vector[i], "_Transcriptome_Profiling_raw_counts.rda")
  file.logic <- grepl(pattern, file.list)
  
  # Compiling path to data 
  data.path <- paste0(data.path, "/", file.list[file.logic])
  
  cat(paste0("For ", data.sets.vector[i], "\n", "File path is:", "\n", data.path))
  
  setwd(prim.dir)
  
  
  # Plot the data
  print(paste0("Plotting", data.sets.vector[i]))
  
  tryCatch({
    DGE.and.OS.plot.function(data.sets.vector[i], 
                           data.loaded = TRUE, 
                           rda.file.path = data.path,
                           genes.to.plot = goi.plot.vector) #goi.plot.vector
    }, error = function(e){cat("\n", "ERROR: ", conditionMessage(e), "\n")})
  
  }



```

